core/domain/exp_domain.py

```diff
@@ -1611,6 +1611,57 @@ def create_default_state(
                 default_dest_state_name))
 
 
+class ExplorationVersionsDiff(object):
+    """Domain object for the difference between two versions of an Oppia
+    exploration.
+
+    Attributes:
+        states_added: list(str). The states added to the exploration from
+            prev_exp_version to current_exp_version.
+        states_deleted: list(str). The states deleted from the exploration from
+            prev_exp_version to current_exp_version.
+        states_renamed: dict. Dictionary mapping state names of
+            current_exp_version to the state names of prev_exp_version.
+    """
+
+    def __init__(self, change_list):
```

Comment:
Don't you also need the state names of the old version of the exploration to seed this (at least as a sanity check that e.g. added names are not already in the list of states, deleted names are already in the list, etc.)?

It would be OK to re-characterize states_renamed as a mapping from states whose "identities" still exist in both the old and new versions, and have a state name map to itself if nothing has happened to it during the change. That way you can keep track of the state changes fully.

Reply:
@seanlip Not necessarily would we need it, would we? For example, we can't use the old state names as a sanity check for deletion of a state since a state can be renamed and then deleted. This sanity check could only work for addition of states and over engineering for this case might not be required right?

Reply:
Re your first paragraph, let's say A gets renamed to B and then gets deleted. I would expect the states_renamed dict to start with A --> A, then become A --> B, and then before deleting B we check that it's either in states_renamed.values() or states_added. I think you do want to take the initial state names into account because otherwise you're kind of assuming things about the exploration based on the changelist. This also acts as a sanity check on our production data.

Re the second paragraph, IMO the efficiency costs in this case would be negligible. I think it's more definitive if you take all states into account since then in the client code that uses this function you won't have to do "if state name is in renamed states then do this". It'd just be "do this".